torch_geometric/nn/conv/message_passing.py

```diff
@@ -94,66 +175,120 @@ def __init__(self, aggr="add", flow="source_to_target", node_dim=0):
         self.__edge_mask__ = None
 
     def __get_mp_type__(self, edge_index):
-        if (torch.is_tensor(edge_index) and edge_index.dtype == torch.long
-                and edge_index.dim() == 2 and edge_index.size(0)):
+        if (isinstance(edge_index, torch.Tensor)
+           and edge_index.dtype == torch.long
+           and edge_index.dim() == 2
+           and edge_index.size(0)):
             return 'edge_index'
         elif isinstance(edge_index, SparseTensor):
             return 'adj_t'
         else:
-            return ValueError(
+            raise ValueError(
                 ('`MessagePassing.propagate` only supports `torch.LongTensor` '
                  'of shape `[2, num_messages]` or `torch_sparse.SparseTensor` '
                  'for argument :obj:`edge_index`.'))
 
-    def __set_size__(self, size, idx, tensor):
-        if not torch.is_tensor(tensor):
+    def __set_size__(self,
+                     size: List[Optional[int]],
+                     idx: int,
+                     tensor: Optional[torch.Tensor]):
+        if not isinstance(tensor, torch.Tensor):
             pass
         elif size[idx] is None:
+            assert tensor is not None
```

Comment:
OK - at some point that was throwing me an error so I just kept it, but the code did change around it. I'll see if I can remove it, thanks!

To answer your questions:
1) kwargs can contain more or less any python type, but it tends to be `float`/`int`/`bool`/`Tensor`, and occasionally `Tuple[Optional[Tensor], Optional[Tensor]]` for bipartite graph stuff.

2) Yeah resolving kwargs on a convop-by-convop basis is the largest reason I needed the codegen, otherwise the jit couldn't handle the type ambiguity. The second bit is dealing better with the various kinds of inputs that can be giving to `MessagePassing.propagate`, it's just a really flexible function and you have to whack it into a shape for the jit to make sense of it.